### PR TITLE
ARROW-18206: [C++][CI] Add a nightly build for C++20 compilation

### PR DESCRIPTION
**Pull request by pitrou**
Fix a couple compilation issues (implicit this-capturing with "=" in lambdas is deprecated in C++20).

### CODE REVIEW

**Comment by pitrou** [comment]
Revision: d2ad465f5cbfcbfb3b7a2da5c07e3cc9e2533f72

Submitted crossbow builds: [ursacomputing/crossbow @ ARROW-18206-cpp20-0](https://github.com/ursacomputing/crossbow/branches/all?query=ARROW-18206-cpp20-0)

|Task|Status|
|----|------|
|test-alpine-linux-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378336262/jobs/5608408512)|
|test-build-cpp-fuzz|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378341384/jobs/5608420148)|
|test-conda-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378343953/jobs/5608426394)|
|test-conda-cpp-valgrind|[Azure](https://github.com/ursacomputing/crossbow/runs/9251115484)|
|test-cuda-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378337119/jobs/5608410364)|
|test-debian-10-cpp-amd64|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378343132/jobs/5608424422)|
|test-debian-10-cpp-i386|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378338396/jobs/5608413369)|
|test-debian-11-cpp-amd64|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378340497/jobs/5608418343)|
|test-debian-11-cpp-i386|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378338026/jobs/5608412551)|
|test-fedora-35-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378340907/jobs/5608419152)|
|test-ubuntu-18.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378342702/jobs/5608423484)|
|test-ubuntu-18.04-cpp-release|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378342229/jobs/5608422237)|
|test-ubuntu-18.04-cpp-static|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378341822/jobs/5608421320)|
|test-ubuntu-20.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378337479/jobs/5608411263)|
|test-ubuntu-20.04-cpp-20|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378340024/jobs/5608417350)|
|test-ubuntu-20.04-cpp-bundled|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378338805/jobs/5608414372)|
|test-ubuntu-20.04-cpp-thread-sanitizer|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378343554/jobs/5608425454)|
|test-ubuntu-22.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378336704/jobs/5608409464)|

**Comment by pitrou** [comment]
Revision: ddfaad532a837529af55ed8d878ccb8b0d0b9d76

Submitted crossbow builds: [ursacomputing/crossbow @ ARROW-18206-cpp20-1](https://github.com/ursacomputing/crossbow/branches/all?query=ARROW-18206-cpp20-1)

|Task|Status|
|----|------|
|test-alpine-linux-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-alpine-linux-cpp)|
|test-build-cpp-fuzz|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378752853/jobs/5609383859)|
|test-conda-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-conda-cpp)|
|test-conda-cpp-valgrind|[Azure](https://github.com/ursacomputing/crossbow/runs/9252414115)|
|test-cuda-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-cuda-cpp)|
|test-debian-10-cpp-amd64|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378752422/jobs/5609382868)|
|test-debian-10-cpp-i386|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-debian-10-cpp-i386)|
|test-debian-11-cpp-amd64|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-debian-11-cpp-amd64)|
|test-debian-11-cpp-i386|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378755026/jobs/5609388649)|
|test-fedora-35-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-fedora-35-cpp)|
|test-ubuntu-18.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378751929/jobs/5609381762)|
|test-ubuntu-18.04-cpp-release|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-ubuntu-18.04-cpp-release)|
|test-ubuntu-18.04-cpp-static|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-ubuntu-18.04-cpp-static)|
|test-ubuntu-20.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378753275/jobs/5609384895)|
|test-ubuntu-20.04-cpp-20|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378754820/jobs/5609388190)|
|test-ubuntu-20.04-cpp-bundled|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-ubuntu-20.04-cpp-bundled)|
|test-ubuntu-20.04-cpp-thread-sanitizer|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3378754139/jobs/5609386876)|
|test-ubuntu-22.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-1-github-test-ubuntu-22.04-cpp)|

**Comment by pitrou** [comment]
Revision: 37a12d00a53a03b3c3f4c4637e65e2cdf31c03c0

Submitted crossbow builds: [ursacomputing/crossbow @ ARROW-18206-cpp20-2](https://github.com/ursacomputing/crossbow/branches/all?query=ARROW-18206-cpp20-2)

|Task|Status|
|----|------|
|test-alpine-linux-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-alpine-linux-cpp)|
|test-build-cpp-fuzz|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-build-cpp-fuzz)|
|test-conda-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380035608/jobs/5612350781)|
|test-conda-cpp-valgrind|[Azure](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-azure-test-conda-cpp-valgrind)|
|test-cuda-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-cuda-cpp)|
|test-debian-10-cpp-amd64|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380034530/jobs/5612348486)|
|test-debian-10-cpp-i386|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-debian-10-cpp-i386)|
|test-debian-11-cpp-amd64|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380034922/jobs/5612349271)|
|test-debian-11-cpp-i386|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380034178/jobs/5612347734)|
|test-fedora-35-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-fedora-35-cpp)|
|test-ubuntu-18.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380036784/jobs/5612353061)|
|test-ubuntu-18.04-cpp-release|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380033859/jobs/5612347133)|
|test-ubuntu-18.04-cpp-static|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-ubuntu-18.04-cpp-static)|
|test-ubuntu-20.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-ubuntu-20.04-cpp)|
|test-ubuntu-20.04-cpp-20|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-ubuntu-20.04-cpp-20)|
|test-ubuntu-20.04-cpp-bundled|[Github Actions](https://github.com/ursacomputing/crossbow//tree/ARROW-18206-cpp20-2-github-test-ubuntu-20.04-cpp-bundled)|
|test-ubuntu-20.04-cpp-thread-sanitizer|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380037073/jobs/5612353915)|
|test-ubuntu-22.04-cpp|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/3380035208/jobs/5612349851)|

File: cpp/src/arrow/compute/exec/source_node.cc [hunk 1]
```diff
@@ -131,22 +131,23 @@ struct SourceNode : ExecNode {
...
-                           if (use_legacy_batching) {
-                             batch_size = morsel_length;
-                           }
-                           ExecBatch batch = morsel.Slice(offset, batch_size);
-                           offset += batch_size;
-                           outputs_[0]->InputReceived(this, std::move(batch));
-                         } while (offset < morsel.length);
-                         return Status::OK();
-                       }));
+                       RETURN_NOT_OK(plan_->ScheduleTask(
```
Comment (pitrou) [inline cpp/src/arrow/compute/exec/source_node.cc:134]: This cascade of nested callbacks is not the most readable code, except perhaps for Lisp fans...

File: cpp/src/arrow/testing/gtest_util.cc [hunk 2]
```diff
@@ -679,7 +679,7 @@ void TestInitialized(const ArrayData& array) {
   // entire buffer data).  If not all bits are well-defined, Valgrind will
   // error with "Conditional jump or move depends on uninitialised value(s)".
   if (total_bit == 0) {
-    ++throw_away;
+    throw_away = throw_away + 1;
```
Comment (westonpace) [inline cpp/src/arrow/testing/gtest_util.cc:682]: Why this change?

**Review by westonpace — APPROVED** [review]
Thanks for working through this.

File: cpp/src/arrow/testing/gtest_util.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/testing/gtest_util.cc:682]: Because C++20 deprecates `++` and `+=` on volatile-qualified variables.
